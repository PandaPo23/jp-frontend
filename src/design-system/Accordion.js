import styled from '@emotion/styled';
import React from 'react';
import { animated, config, useSpring } from 'react-spring';

import Box from './Box';
import Flex from './Flex';
import FlexPanel from './FlexPanel';
import { Heading6 } from './Heading';
import Icon from './Icon';

const Accordion = ({
  children,
  openedSection,
  name,
  animated = true,
  onSectionToggle = () => {},
}) => (
  <FlexPanel>
    {React.Children.map(children, (child) =>
      React.cloneElement(child, {
        onSectionToggle: () =>
          onSectionToggle({ accordion: name, section: child.props.name }),
        open: openedSection === child.props.name,
        animated,
      })
    )}
  </FlexPanel>
);

const AccordionSection = ({
  children,
  open = false,
  disabled = false,
  name,
  animated,
  onSectionToggle,
  ...props
}) => {
  const childrenArray = Array.isArray(children) ? children : [children];
  return (
    <Flex
      flexDirection="column"
      name={name}
      disabled={disabled}
      open={open}
      borderBottom={2}
      borderColor="selects.tagBorder"
      {...props}
    >
      {childrenArray.map((child, index) =>
        child.type === AccordionHeading
          ? React.cloneElement(child, {
              key: child.key || index,
              onClick: onSectionToggle,
              open,
              disabled,
            })
          : React.cloneElement(child, {
              key: child.key || index,
              open,
              animated,
            })
      )}
    </Flex>
  );
};

/**
 *
 * @param {*} visibleWhen
 * @param {*} open
 * @return {bool}
 */
const isVisible = (visibleWhen, open) => {
  switch (visibleWhen) {
    case 'expanded':
      return open;
    case 'collapsed':
      return !open;
    case 'either':
    default:
      return true;
  }
};

/**
 * Handle icons rendering
 * @param {*} iconString
 * @param {*} open
 * @return {string}
 */
const getIcon = (iconString, open) => {
  const icons = iconString.split(/[, ]/);
  const idx = icons.length === 1 ? 0 : open ? 1 : 0;
  return icons[idx];
};

const AccordionHeading = ({
  iconPosition = 'right',
  icon = 'plus,minus',
  iconSize = 16,
  children,
  onClick,
  open,
  disabled,
  visibleWhen = 'either',
  ...props
}) =>
  isVisible(visibleWhen, open) ? (
    <Flex
      alignItems="center"
      hoverable={!disabled}
      pl={4}
      pr={3}
      justifyContent={iconPosition === 'right' && 'space-between'}
      onClick={!disabled ? onClick : null}
      {...props}
    >
      {iconPosition === 'left' && (
        <Icon
          size={iconSize}
          name={getIcon(icon, open)}
          color={disabled ? 'on.disabled' : undefined}
        />
      )}
      <Heading6
        pl={iconPosition === 'left' ? 3 : undefined}
        pr={iconPosition === 'right' ? 3 : undefined}
        py={3}
        uppercase
        color={disabled ? 'on.disabled' : undefined}
        fontWeight="normal"
      >
        {children}
      </Heading6>
      {iconPosition === 'right' && (
        <Icon
          size={iconSize}
          name={getIcon(icon, open)}
          color={disabled ? 'on.disabled' : undefined}
        />
      )}
    </Flex>
  ) : null;

const AnimatedPanel = styled(animated.div)`
  overflow-y: auto;
`;

const AccordionContent = ({ children, open, maxHeight = 600, animated }) => {
  let { height } = useSpring({
    from: { height: open ? 0 : maxHeight },
    height: open ? maxHeight : 0,
    config: config.fast,
    immediate: !animated,
  });
  return (
    <AnimatedPanel style={{ height }}>
      <Box p={4}>{children}</Box>
    </AnimatedPanel>
  );
};

Accordion.Section = AccordionSection;
Accordion.Heading = AccordionHeading;
Accordion.Content = AccordionContent;

export default Accordion;
