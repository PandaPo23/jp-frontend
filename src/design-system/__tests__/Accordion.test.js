import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import mockDate from '../../__mocks__/dateMocks';

import Accordion from '../Accordion';

describe('Accordion component', () => {
  let timestamp;
  const getComponent = (props) => (
    <Accordion {...props}>
      <Accordion.Section name="section1">
        <Accordion.Heading visibleWhen="collapsed">
          Collapsed Heading 1
        </Accordion.Heading>
        <Accordion.Heading visibleWhen="expanded">
          Expanded Heading 1
        </Accordion.Heading>
        <Accordion.Content>Accordion Content 2</Accordion.Content>
      </Accordion.Section>
      <Accordion.Section name="section2">
        <Accordion.Heading visibleWhen="collapsed">
          Collapsed Heading 2
        </Accordion.Heading>
        <Accordion.Heading visibleWhen="expanded">
          Expanded Heading 2
        </Accordion.Heading>
        <Accordion.Content>Accordion Content 2</Accordion.Content>
      </Accordion.Section>
    </Accordion>
  );

  beforeEach(() => {
    timestamp = mockDate();
  });

  it('renders accordion correctly', () => {
    const component = getComponent({
      name: 'Accordion',
      onSectionToggle: () => {},
      openedSection: 'section2',
    });
    const wrapper = mount(component);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders the content of active section only', () => {
    const component = getComponent({
      name: 'Accordion',
      onSectionToggle: () => {},
      openedSection: 'section2',
    });
    const wrapper = mount(component);
    expect(
      wrapper
        .find(Accordion.Section)
        .at(0)
        .find(Accordion.Content)
        .prop('open')
    ).toBe(false);
    expect(
      wrapper
        .find(Accordion.Section)
        .at(1)
        .find(Accordion.Content)
        .prop('open')
    ).toBe(true);
  });

  it('triggers `onSectionToggle` when section header is clicked', () => {
    const accordionProps = {
      name: 'Accordion',
      onSectionToggle: jest.fn(),
      openedSection: 'section2',
    };
    const component = getComponent(accordionProps);
    const wrapper = mount(component);
    wrapper
      .find(Accordion.Section)
      .first()
      .find(Accordion.Heading)
      .first()
      .simulate('click');
    expect(accordionProps.onSectionToggle).toHaveBeenCalled();
  });
});
