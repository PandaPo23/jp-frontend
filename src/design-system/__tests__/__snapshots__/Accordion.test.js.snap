// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion component renders accordion correctly 1`] = `
.emotion-20 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-height: 0;
}

.emotion-9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border-bottom: 2px;
  border-color: selects.tagBorder;
}

.emotion-5 {
  padding-left: 32px;
  padding-right: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
}

.emotion-0 {
  padding-top: 16px;
  padding-bottom: 16px;
  padding-right: 16px;
  font-weight: normal;
  text-transform: uppercase;
}

.emotion-1 {
  min-width: 16px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  min-width: 16px;
}

.emotion-7 {
  overflow-y: auto;
}

.emotion-6 {
  padding: 32px;
}

<Accordion
  name="Accordion"
  onSectionToggle={[Function]}
  openedSection="section2"
>
  <FlexPanel
    display="flex"
    flexDirection="column"
  >
    <div
      className="emotion-20"
    >
      <AccordionSection
        animated={true}
        key=".0"
        name="section1"
        onSectionToggle={[Function]}
        open={false}
      >
        <Flex
          borderBottom={2}
          borderColor="selects.tagBorder"
          disabled={false}
          display="flex"
          flexDirection="column"
          name="section1"
          open={false}
        >
          <div
            className="emotion-9"
            disabled={false}
            name="section1"
            open={false}
          >
            <AccordionHeading
              disabled={false}
              key="0"
              onClick={[Function]}
              open={false}
              visibleWhen="collapsed"
            >
              <Flex
                alignItems="center"
                display="flex"
                hoverable={true}
                justifyContent="space-between"
                onClick={[Function]}
                pl={4}
                pr={3}
              >
                <div
                  className="emotion-5"
                  onClick={[Function]}
                >
                  <Heading6
                    as="h6"
                    className="heading"
                    fontWeight="normal"
                    pr={3}
                    py={3}
                    uppercase={true}
                  >
                    <h6
                      className="heading emotion-0"
                    >
                      Collapsed Heading 1
                    </h6>
                  </Heading6>
                  <Icon
                    name="plus"
                    size={16}
                  >
                    <Base
                      className="emotion-1"
                      name="plus"
                      size={16}
                    >
                      <Flex
                        centered={true}
                        className="emotion-1"
                        display="flex"
                      >
                        <div
                          className="emotion-2"
                        >
                          <svg
                            className="emotion-1"
                            name="plus"
                            size={16}
                          />
                        </div>
                      </Flex>
                    </Base>
                  </Icon>
                </div>
              </Flex>
            </AccordionHeading>
            <AccordionHeading
              disabled={false}
              key="1"
              onClick={[Function]}
              open={false}
              visibleWhen="expanded"
            />
            <AccordionContent
              animated={true}
              key="2"
              open={false}
            >
              <Styled(Component)
                style={
                  Object {
                    "height": AnimatedValue {
                      "animatedStyles": Set {},
                      "children": Array [
                        AnimatedStyle {
                          "attach": [Function],
                          "children": Array [
                            AnimatedProps {
                              "attach": [Function],
                              "children": Array [],
                              "detach": [Function],
                              "payload": Object {
                                "children": <ForwardRef(render)
                                  p={4}
                                >
                                  Accordion Content 2
                                </ForwardRef(render)>,
                                "className": "emotion-7",
                                "style": [Circular],
                              },
                              "update": [Function],
                            },
                          ],
                          "detach": [Function],
                          "payload": [Circular],
                        },
                      ],
                      "done": false,
                      "lastPosition": 600,
                      "lastTime": undefined,
                      "lastVelocity": undefined,
                      "payload": undefined,
                      "setValue": [Function],
                      "startPosition": 600,
                      "startTime": 1479427200000,
                      "value": 600,
                    },
                  }
                }
              >
                <ForwardRef
                  className="emotion-7"
                  style={
                    Object {
                      "height": AnimatedValue {
                        "animatedStyles": Set {},
                        "children": Array [
                          AnimatedStyle {
                            "attach": [Function],
                            "children": Array [
                              AnimatedProps {
                                "attach": [Function],
                                "children": Array [],
                                "detach": [Function],
                                "payload": Object {
                                  "children": <ForwardRef(render)
                                    p={4}
                                  >
                                    Accordion Content 2
                                  </ForwardRef(render)>,
                                  "className": "emotion-7",
                                  "style": [Circular],
                                },
                                "update": [Function],
                              },
                            ],
                            "detach": [Function],
                            "payload": [Circular],
                          },
                        ],
                        "done": false,
                        "lastPosition": 600,
                        "lastTime": undefined,
                        "lastVelocity": undefined,
                        "payload": undefined,
                        "setValue": [Function],
                        "startPosition": 600,
                        "startTime": 1479427200000,
                        "value": 600,
                      },
                    }
                  }
                >
                  <div
                    className="emotion-7"
                    style={
                      Object {
                        "height": 600,
                      }
                    }
                  >
                    <Box
                      p={4}
                    >
                      <div
                        className="emotion-6"
                      >
                        Accordion Content 2
                      </div>
                    </Box>
                  </div>
                </ForwardRef>
              </Styled(Component)>
            </AccordionContent>
          </div>
        </Flex>
      </AccordionSection>
      <AccordionSection
        animated={true}
        key=".1"
        name="section2"
        onSectionToggle={[Function]}
        open={true}
      >
        <Flex
          borderBottom={2}
          borderColor="selects.tagBorder"
          disabled={false}
          display="flex"
          flexDirection="column"
          name="section2"
          open={true}
        >
          <div
            className="emotion-9"
            disabled={false}
            name="section2"
            open={true}
          >
            <AccordionHeading
              disabled={false}
              key="0"
              onClick={[Function]}
              open={true}
              visibleWhen="collapsed"
            />
            <AccordionHeading
              disabled={false}
              key="1"
              onClick={[Function]}
              open={true}
              visibleWhen="expanded"
            >
              <Flex
                alignItems="center"
                display="flex"
                hoverable={true}
                justifyContent="space-between"
                onClick={[Function]}
                pl={4}
                pr={3}
              >
                <div
                  className="emotion-5"
                  onClick={[Function]}
                >
                  <Heading6
                    as="h6"
                    className="heading"
                    fontWeight="normal"
                    pr={3}
                    py={3}
                    uppercase={true}
                  >
                    <h6
                      className="heading emotion-0"
                    >
                      Expanded Heading 2
                    </h6>
                  </Heading6>
                  <Icon
                    name="minus"
                    size={16}
                  >
                    <Base
                      className="emotion-1"
                      name="minus"
                      size={16}
                    >
                      <Flex
                        centered={true}
                        className="emotion-1"
                        display="flex"
                      >
                        <div
                          className="emotion-2"
                        >
                          <svg
                            className="emotion-1"
                            name="minus"
                            size={16}
                          />
                        </div>
                      </Flex>
                    </Base>
                  </Icon>
                </div>
              </Flex>
            </AccordionHeading>
            <AccordionContent
              animated={true}
              key="2"
              open={true}
            >
              <Styled(Component)
                style={
                  Object {
                    "height": AnimatedValue {
                      "animatedStyles": Set {},
                      "children": Array [
                        AnimatedStyle {
                          "attach": [Function],
                          "children": Array [
                            AnimatedProps {
                              "attach": [Function],
                              "children": Array [],
                              "detach": [Function],
                              "payload": Object {
                                "children": <ForwardRef(render)
                                  p={4}
                                >
                                  Accordion Content 2
                                </ForwardRef(render)>,
                                "className": "emotion-7",
                                "style": [Circular],
                              },
                              "update": [Function],
                            },
                          ],
                          "detach": [Function],
                          "payload": [Circular],
                        },
                      ],
                      "done": false,
                      "lastPosition": 0,
                      "lastTime": undefined,
                      "lastVelocity": undefined,
                      "payload": undefined,
                      "setValue": [Function],
                      "startPosition": 0,
                      "startTime": 1479427200000,
                      "value": 0,
                    },
                  }
                }
              >
                <ForwardRef
                  className="emotion-7"
                  style={
                    Object {
                      "height": AnimatedValue {
                        "animatedStyles": Set {},
                        "children": Array [
                          AnimatedStyle {
                            "attach": [Function],
                            "children": Array [
                              AnimatedProps {
                                "attach": [Function],
                                "children": Array [],
                                "detach": [Function],
                                "payload": Object {
                                  "children": <ForwardRef(render)
                                    p={4}
                                  >
                                    Accordion Content 2
                                  </ForwardRef(render)>,
                                  "className": "emotion-7",
                                  "style": [Circular],
                                },
                                "update": [Function],
                              },
                            ],
                            "detach": [Function],
                            "payload": [Circular],
                          },
                        ],
                        "done": false,
                        "lastPosition": 0,
                        "lastTime": undefined,
                        "lastVelocity": undefined,
                        "payload": undefined,
                        "setValue": [Function],
                        "startPosition": 0,
                        "startTime": 1479427200000,
                        "value": 0,
                      },
                    }
                  }
                >
                  <div
                    className="emotion-7"
                    style={
                      Object {
                        "height": 0,
                      }
                    }
                  >
                    <Box
                      p={4}
                    >
                      <div
                        className="emotion-6"
                      >
                        Accordion Content 2
                      </div>
                    </Box>
                  </div>
                </ForwardRef>
              </Styled(Component)>
            </AccordionContent>
          </div>
        </Flex>
      </AccordionSection>
    </div>
  </FlexPanel>
</Accordion>
`;
